yellowbrick/cluster/silhouette.py

```diff
@@ -135,18 +135,30 @@ def fit(self, X, y=None, **kwargs):
         # NOTE: Probably this would be better in score, but the standard score
         # is a little different and I'm not sure how it's used.
 
-        if not check_fitted(self.estimator, is_fitted_by=self.is_fitted):
-            # Fit the wrapped estimator
-            self.estimator.fit(X, y, **kwargs)
+        # if estimator is fitted AND has attribute
+        if check_fitted(self.estimator, is_fitted_by=self.is_fitted) and hasattr(self.estimator, "predict"):
+            labels = self.estimator.predict(X)
+        else:  # if estimator is NOT fitted, OR estimator does NOT implement predict()
+            labels = self.estimator.fit_predict(X, y, **kwargs)
 
         # Get the properties of the dataset
         self.n_samples_ = X.shape[0]
-        self.n_clusters_ = self.estimator.n_clusters
+
+        if hasattr(self.estimator, "n_clusters"):
+            self.n_clusters_ = self.estimator.n_clusters
+        else:
+            self.n_clusters_ = len(set(labels)) - (1 if -1 in labels else 0)
```

Comment:
I wonder if there is a more expressive way to write this (specifically the 1 if -1 in labels else 0)

Reply:
```suggestion
            unique_labels = set(labels)
            n_noise_clusters = 1 if -1 in unique_labels else 0
            self.n_clusters_ = len(unique_labels) - n_noise_clusters 
```

Is this any better?